scripts/test-vanilla-compliance

```diff
@@ -0,0 +1,642 @@
+            ". Not fixable.")
+
+    return (bvert, evert, flags, linet, stag, rsdef, lsdef)
+
+
+# Process (do) a region. A region can be either a lump, or some other non-lump
+# array of structures (the directory). Note that unlike the other do_*
+# methods this method process the entire region, not just one item in it.
+def do_region(region_name, map_name, wad, fhand, lump_offset, lump_size):
+    info = region_info.get(region_name)
+    if info is None:
+        # Nothing is supported for this region type.
+        return 0
+    old_pos = fhand.tell()
+    fhand.seek(lump_offset)
+    unpack_format, field_names_part, row_format_part, do_item = info
+    unpack_size = struct.calcsize(unpack_format)
+    field_names = ("Region", "Map", "Num") + field_names_part + (
+        " Fixes" if args.fix else " Errors",)
+    row_format = "%" + str(len(region_name)) + "s %5s %5d " + \
+        row_format_part + "%s"
+    header_format = row_format.replace("d", "s")
+    header = header_format % field_names
+
+    count = 0 # Count of errors or fixes.
+    offset = 0
+    num = 0
+    first = len(region_header_shown) == 0
+    while offset < lump_size:
+        bytes = fhand.read(unpack_size)
+        fields = wad_unpack(unpack_format, bytes)
+        messages = []
+        fields = do_item(messages, num, *fields)
+        if args.fix and len(messages) > 0:
+            fhand.seek(fhand.tell() - unpack_size)
+            bytes = wad_pack(unpack_format, *fields)
+            fhand.write(bytes)
+        if args.verbose or len(messages) > 0:
+            if not region_name in region_header_shown:
+                region_header_shown[region_name] = set()
+            shown = region_header_shown[region_name]
+            if not wad in shown:
+                if not first:
+                    print("")
+                print("Process", region_name, "for", map_name, "in", wad)
+                print(header)
+                shown.add(wad)
+            row = (region_name, map_name, num) + fields + (
+                get_combined_message(messages),)
+            print(row_format % row)
+            count += len(messages)
+        offset += unpack_size
+        num += 1
+    fhand.seek(old_pos)
+    return count
+
+
+# Process a SECTORS item.
+def do_sector(messages, number, fhght, chght, ftext, ctext, light, typ, tag):
+    # There's no obvious automated way of fixing sector types.
+    if (not (typ in allowed_sectors_types)) and not args.fix:
+        messages.append("Type " + str(typ) + " is not allowed and can not "
+            "be fixed.")
+
+    return (fhght, chght, ftext, ctext, light, typ, tag)
+
+
+# Process a THINGS item.
+def do_thing(messages, number, x, y, angle, typ, spawn_flags):
+    # In the following sections each check one field within the thing. They are
+    # in file order.
+
+    # Make sure that the angle is in range [0, 360) and that it is a
+    # multiple of 45. The "//" rounds down, so the " + 22" makes it round
```

Comment:
Thanks for your helpful feedback. I'll have more time to go through and apply your feedback this weekend, but for now I thought I'd elaborate on my motivation and thinking just in case in it's interesting, or has any impact.

The multiple of 45° is implied in a few places that I've seen, but I haven't found anywhere that it's required. Some level editors I've seen make it easy to select a multiple of 45°, but also allow arbitrary angles to be entered.

Id IWADs doom.wad, doom2.wad, plutonia.wad and tnt.wad overwhelming choose multiples of 45° with just a few possible mistakes, but you're correct that it won't crash (at least as far as I know).

Since you've looked at more Doom engine code than I have, do you happen to know if different source ports are likely to handle the non-multiple of 45° differently, some rounding, some using it accurately, etc.? If so that seems like a (at least theoretical) testing burden that could be avoided.

There's the less tangible argument that it would be nice if the Freedoom IWADs were exemplary of old school IWADs even if it does not impact anything. People without access to the proprietary IWADs could get an idea of what IWADs were, or at least tried to be, like.

This conversation was about angles, but it could also be applied to other field cleanup that isn't obviously helpful, which you mentioned in other comments.

As to burdening non-programmer map creators - I was picturing this cleanup as being an effort parallel to the mapping work, possibly by different people. I was also assuming that the maps are stable to various degrees in Freedoom with many of them stable enough that they are unlikely to have additional changes, and therefore are good candidates for this cleanup. I could however soften the language I added in `README.adoc` to make it more of a suggestion, to make it clear that it's ok if it's not done. It can be done eventually once the map is more stable.

Also, if you'd rather discuss this on the forum I created a thread [here](https://www.doomworld.com/forum/topic/122490-my-pull-request-for-my-test-vanilla-compliance-script/).

Reply:
Angle checking is now off by default.